Add a Get Started doc with three scenarios (#373)
open-telemetry/opentelemetry-go-instrumentation: docs/get-started.md

Reviewer: How about moving this to the main `README.md`?

Author: Let's!

# Instrument a Go application or services automatically

You can instrument a Go executable using OpenTelemetry without having
to write additional code. All you need to do is configure a few environment
variables and run the instrumentation with elevated privileges.

The following example shows how to instrument a simple Go application
automatically on a Linux host, through Docker, and using Kubernetes.

## Prerequisites

To instrument an application automatically, you need the following:

- Binary file or Docker image of OpenTelemetry Go instrumentation

> - Go 1.17 and higher in your system

Reviewer 2: I'm not sure you need Go installed if you already have a binary or docker image of the auto-instrumentation, right?

If you want to build the binary it will require 1.18, not 1.17, though.

Author: @pellared We're not offering a binary of the instrumentation yet thought, right?

Reviewer 2: Right, we are not offering a binary for download yet. Only a docker image.

Maybe we want to say you will need the docker image or the source code and at least Go 1.18 to compile a binary?

> - Linux with kernel version 4.4 and higher

Reviewer 2: eBPF also needs to be loaded on the system.

Author: Oh, I though it was on by default.

Author: @pellared How do you enable it? Is this only for kernels of the 4.x family?

Reviewer: I think it should be enabled by default 🤷

Reviewer: @edeNFed any insights?

Reviewer 2: Yeah, looking more at this, the default compilation will have eBPF enabled. I think this can be resolved.

- x64 or ARM processor

## Instrument an application on the same host

To instrument an application on the same host, follow these steps:

1. Set the following environment variables:

  - `OTEL_GO_AUTO_TARGET_EXE`: Full path of the executable you want to
  instrument
  - `OTEL_EXPORTER_OTLP_ENDPOINT`: Your observability backend. For example,
  `localhost:4317`
  - `OTEL_SERVICE_NAME`: Name of your service or application

2. Run the target application.

3. Run the OpenTelemetry Go instrumentation with root privileges.

> ## Instrument an application in Docker

Reviewer 2: ```suggestion
## Instrument an application in Docker using docker-compose
```

Reviewer: ```suggestion
## Instrument an application in Docker Compose
```

To instrument a containerized application, follow these steps:

1. Create or edit the docker-compose.yaml file. Make sure to add a Docker
network , a shared volume, and a service for the application  

2. Edit the docker-compose file to add a new service for the instrumentation:

  ```yaml
    go-auto:
      depends_on:
        - <name_of_your_application_service>

>       build:
>         context: <relative_path_to_instrumentation_Dockerfile>
>         dockerfile: Dockerfile

Reviewer: I think it should be

```suggestion
      image: otel/autoinstrumentation-go
```

      privileged: true
      pid: "host"
      environment:
        - OTEL_EXPORTER_OTLP_ENDPOINT=http://<address_in_docker_network>:4317
        - OTEL_GO_AUTO_TARGET_EXE=<location_of_target_application_binary>
        - OTEL_SERVICE_NAME=<name_of_your_application>
        - OTEL_PROPAGATORS=tracecontext,baggage
      volumes:
        - <shared_volume_of_application>
        - /proc:/host/proc
  ```

3. Run `docker compose up`.

## Instrument an application in Kubernetes

To instrument an application running in Kubernetes, follow these steps:

1. Add the container image of the OpenTelemetry Go instrumentation to your manifest. Make sure that `runAsUser` is set to `0`, `privileged` is set to `true`, and that the `SYS_PTRACE` capability is present:

   ```yaml
   - name: <your_application_name>

>      image: otel-go-instrumentation

Reviewer: ```suggestion
     image: otel/autoinstrumentation-go
```

     imagePullPolicy: IfNotPresent
     # env variables go here
     securityContext:
       runAsUser: 0
       capabilities:
         add:
           - SYS_PTRACE
       privileged: true
   ```

> 2. Add the OTel Go environment variables. For example:

Reviewer: How about merging the steps as you did for docker compose?

   ```yaml
   - name: <your_application_name>

>      image: otel-go-instrumentation

Reviewer: ```suggestion
     image: otel/autoinstrumentation-go
```

      terminationGracePeriodSeconds: 0
      containers:
        - env:
            - name: WEB_PORT
              value: "8080"
            - name: EMOJISVC_HOST
              value: emoji-svc.emojivoto:8080
            - name: VOTINGSVC_HOST
              value: voting-svc.emojivoto:8080
            - name: INDEX_BUNDLE
              value: dist/index_bundle.js
          image: docker.l5d.io/buoyantio/emojivoto-web:v11
          name: web-svc
          ports:
            - containerPort: 8080
              name: http
          resources:
            requests:
              cpu: 100m
        - name: <your_application_name>
          image: otel-go-instrumentation
          imagePullPolicy: IfNotPresent
          env:
            - name: OTEL_GO_AUTO_TARGET_EXE
              value: <location_of_target_application_binary>
            - name: OTEL_EXPORTER_OTLP_ENDPOINT
              value: "http://<address_in_network>:4317"
            - name: OTEL_SERVICE_NAME
              value: "<name_of_service>"
          securityContext:
            runAsUser: 0
            capabilities:
              add:
                - SYS_PTRACE
            privileged: true

> ```

Author: I'm unsure as well... Thought about providing a complete example, but perhaps it's too much handholding.

> ```

Reviewer: I think this is too much especially that there is nothing like that e.g. for docker compose.

- x64 or ARM processor
- Docker image or compiled binary of OpenTelemetry Go Automatic Instrumentation

To compile the instrumentation binary, use Go 1.18 or higher.

## Instrument an application on the same host

To instrument an application on the same host, follow these steps:

1. Set the following environment variables:

  - `OTEL_GO_AUTO_TARGET_EXE`: Full path of the executable you want to
  instrument. For example, `/home/bin/service_executable`
  - `OTEL_SERVICE_NAME`: Name of your service or application
  - `OTEL_EXPORTER_OTLP_ENDPOINT`: Your observability backend. For example,
  `http://localhost:4317. If you're sending data to the OpenTelemetry Collector
  over a secure gRPC transport, set TLS settings in the OTLP exporter. See the
  [OTLP Receiver documentation](https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/otlpreceiver/README.md)
  for instructions

2. Run the target application.

3. Run the OpenTelemetry Go Automatic Instrumentation with root privileges.

## Instrument an application in Docker Compose

To instrument a containerized application, follow these steps:

1. Create or edit the docker-compose.yaml file. Make sure to add a Docker

> network , a shared volume, and a service for the application.

Reviewer: ```suggestion
network  a shared volume, and a service for the application.
```

# Instrument a Go application or services automatically

You can instrument a Go executable using OpenTelemetry without having
to write additional code. All you need to do is configure a few environment
variables and run the instrumentation with elevated privileges.

The following example shows how to instrument a simple Go application
automatically on a Linux host, through Docker, and using Kubernetes.

## Prerequisites

To instrument an application automatically, you need the following:

- Linux with kernel version 4.19 or higher
- x64 or ARM processor

> - Docker image or compiled binary of OpenTelemetry Go instrumentation

Reviewer: ```suggestion
- Docker image or compiled binary of OpenTelemetry Go Automatic Instrumentation
```

To compile the instrumentation binary, use Go 1.18 or higher.

## Instrument an application on the same host

To instrument an application on the same host, follow these steps:

1. Set the following environment variables:

  - `OTEL_GO_AUTO_TARGET_EXE`: Full path of the executable you want to
  instrument. For example, `/home/bin/service_executable`
  - `OTEL_SERVICE_NAME`: Name of your service or application
  - `OTEL_EXPORTER_OTLP_ENDPOINT`: Your observability backend. For example,
  `http://localhost:4317. If you're sending data to the OpenTelemetry Collector

>   over HTTPS, set TLS settings in the OTLP exporter. See the

Reviewer: ```suggestion
  over a secure gRPC transport, set TLS settings in the OTLP exporter. See the
```

  [OTLP Receiver documentation](https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/otlpreceiver/README.md)

>   for instructions

Reviewer: These env vars are only needed for the OpenTelemetry Go Automatic Instrumentation process. I suggest to move this to the "Run the OpenTelemetry Go instrumentation with root privileges." step. E.g.

```
Run the OpenTelemetry Go Automatic Instrumentation with root privileges and following environment variables:
```

2. Run the target application.

> 3. Run the OpenTelemetry Go instrumentation with root privileges.

Reviewer: ```suggestion
3. Run the OpenTelemetry Go Automatic Instrumentation with root privileges.
```

## Instrument an application in Docker Compose

To instrument a containerized application, follow these steps:

1. Create or edit the docker-compose.yaml file. Make sure to add a Docker
network , a shared volume, and a service for the application.

2. Edit the docker-compose file to add a new service for the instrumentation:

  ```yaml
    go-auto:
      depends_on:
        - <name_of_your_application_service>
      image: otel/autoinstrumentation-go
      privileged: true

>       cap_add:
>         - SYS_PTRACE

Reviewer: ```suggestion
```

not needed

      pid: "host"
      environment:
        - OTEL_EXPORTER_OTLP_ENDPOINT=http://<address_in_docker_network>:4317
        - OTEL_GO_AUTO_TARGET_EXE=<location_of_target_application_binary>
        - OTEL_SERVICE_NAME=<name_of_your_application>
        - OTEL_PROPAGATORS=tracecontext,baggage
      volumes:
        - <shared_volume_of_application>
        - /proc:/host/proc
  ```

3. Run `docker compose up`.

## Instrument an application in Kubernetes

To instrument an application running in Kubernetes, follow these steps:

> 1. Add the container image of the OpenTelemetry Go instrumentation to your manifest. Make sure that `runAsUser` is set to `0`, `privileged` is set to `true`, and that the `SYS_PTRACE` capability is present:

Reviewer: ```suggestion
1. Add the container image of the OpenTelemetry Go Automatic Instrumentation to your manifest. Make sure that `runAsUser` is set to `0`, `privileged` is set to `true`:
```

   ```yaml

>    - name: <your_application_name>

Reviewer: ```suggestion
   - name: autoinstrumentation-go
```
